File: lab4/src/container/MapFlatContainer.java
package container;

import java.util.ArrayList;
import java.util.List;
import java.util.TreeMap;
import java.util.stream.Collectors;
import data.Flat;

public class MapFlatContainer implements FlatsContainer {

    private final TreeMap<Double, List<Flat>> flats;

    public MapFlatContainer(List<Flat> flats) {
        this.flats = new TreeMap<Double, List<Flat>>();

        for (Flat flat : flats) {
            add(flat);
        }
    }

    @Override
    public List<Flat> findByArea(double lowerArea, double upperArea) {
        var flatsByArea = this.flats.subMap(lowerArea, true, upperArea, true);

        return flatsByArea.entrySet().stream()
                .flatMap(s -> s.getValue().stream())
                .collect(Collectors.toList());
    }

    @Override
    public void add(Flat flat) {
        var area = flat.getArea();

        List<Flat> currentFlats;
        if (this.flats.containsKey(flat.getArea())) {
            currentFlats = flats.get(area);
        } else {
            currentFlats = new ArrayList<Flat>();
        }

        currentFlats.add(flat);

        this.flats.put(area, currentFlats);
    }

    @Override
    public String toString() {
        return this.flats.entrySet().stream()
                .flatMap(s -> s.getValue().stream())
                .map(Object::toString)
                .collect(Collectors.joining("\n"));
    }
}

File: lab6/src/container/MapFlatContainer.kt
package container

import data.Flat
import java.util.*

class MapFlatContainer(flats: List<Flat>) : FlatsContainer {

    private val flatsMap: TreeMap<Double, MutableList<Flat>> = TreeMap()

    init {
        flats.forEach { add(it) }
    }

    override fun findByArea(lowerArea: Double, upperArea: Double): List<Flat> {
        val flatsByArea = flatsMap.subMap(lowerArea, true, upperArea, true)

        return flatsByArea.values.flatten()
    }

    override fun add(flat: Flat) {
        val area = flat.area

        val currentFlats = flatsMap.getOrPut(area) { mutableListOf() }
        currentFlats.add(flat)
    }

    override fun toString(): String {
        return flatsMap.values.flatten().joinToString("\n") { it.toString() }
    }
}
